File: src/python/agent_builder.py
from __future__ import annotations

from typing import Any, Callable, Self

from src.python.has_hooks import HookEvent
from src.python.has_middleware import BaseMiddleware
from src.python.uses_tools import ToolDef
from src.python.has_skills import Skill
from src.python.event_stream_parser import EventType


class AgentBuilder:
    """Declarative builder for StandardAgent."""

    def __init__(self, model: str) -> None:
        self._model = model
        self._system: str | None = None
        self._max_turns: int = 20
        self._max_retries: int = 2
        self._stream: bool = True
        self._litellm_kwargs: dict[str, Any] = {}
        self._tools: list[Callable | ToolDef] = []
        self._middleware: list[BaseMiddleware] = []
        self._hooks: list[tuple[HookEvent, Callable]] = []
        self._events: list[EventType] = []
        self._skills: list[tuple[Skill, dict[str, Any] | None]] = []
        self._shell_opts: dict[str, Any] | None = None
        self._driver: str | None = None
        self._commands: list[tuple[str, Callable]] = []

    def system(self, prompt: str) -> Self:
        self._system = prompt
        return self

    def max_turns(self, n: int) -> Self:
        self._max_turns = n
        return self

    def max_retries(self, n: int) -> Self:
        self._max_retries = n
        return self

    def stream(self, enabled: bool = True) -> Self:
        self._stream = enabled
        return self

    def litellm(self, **kwargs: Any) -> Self:
        self._litellm_kwargs.update(kwargs)
        return self

    def tool(self, fn_or_def: Callable | ToolDef) -> Self:
        self._tools.append(fn_or_def)
        return self

    def tools(self, *fns_or_defs: Callable | ToolDef) -> Self:
        self._tools.extend(fns_or_defs)
        return self

    def middleware(self, *mws: BaseMiddleware) -> Self:
        self._middleware.extend(mws)
        return self

    def on(self, event: HookEvent, callback: Callable) -> Self:
        self._hooks.append((event, callback))
        return self

    def event(self, event_type: EventType) -> Self:
        self._events.append(event_type)
        return self

    def events(self, *event_types: EventType) -> Self:
        self._events.extend(event_types)
        return self

    def skill(self, sk: Skill, config: dict[str, Any] | None = None) -> Self:
        self._skills.append((sk, config))
        return self

    def skills(self, *sks: Skill) -> Self:
        for sk in sks:
            self._skills.append((sk, None))
        return self

    def shell(self, **kwargs: Any) -> Self:
        self._shell_opts = kwargs
        return self

    def driver(self, name: str) -> Self:
        self._driver = name
        return self

    def command(self, name: str, handler: Callable) -> Self:
        self._commands.append((name, handler))
        return self

    async def create(self) -> Any:
        from src.python.standard_agent import StandardAgent

        agent = StandardAgent(
            model=self._model,
            system=self._system,
            max_turns=self._max_turns,
            max_retries=self._max_retries,
            stream=self._stream,
            **self._litellm_kwargs,
        )

        for fn_or_def in self._tools:
            agent.register_tool(fn_or_def)

        for mw in self._middleware:
            agent.use(mw)

        for event, cb in self._hooks:
            agent.on(event, cb)

        for et in self._events:
            agent.register_event(et)

        if self._shell_opts is not None:
            agent.__init_has_shell__(driver=self._driver, **self._shell_opts)
        elif self._driver is not None:
            agent.__init_has_shell__(driver=self._driver)

        for name, handler in self._commands:
            agent.register_command(name, handler)

        # Skills last (async mount)
        for sk, config in self._skills:
            await agent.mount(sk, config)

        return agent

File: src/typescript/agent-builder.ts
import { HookEvent } from "./has-hooks.js";
import type { Middleware } from "./has-middleware.js";
import type { ToolDef } from "./uses-tools.js";
import type { StructuredEvent } from "./event-stream-parser.js";
import type { Skill } from "./has-skills.js";
import type { CmdHandler } from "./shell.js";
import type { HasShellOptions } from "./has-shell.js";

export class AgentBuilder {
  private _model: string;
  private _system?: string;
  private _maxTurns: number = 20;
  private _maxRetries: number = 2;
  private _stream: boolean = true;
  private _tools: ToolDef[] = [];
  private _middleware: Middleware[] = [];
  private _hooks: Array<[HookEvent, (...args: any[]) => any]> = [];
  private _events: StructuredEvent[] = [];
  private _skills: Array<[Skill, Record<string, any>?]> = [];
  private _shellOpts?: HasShellOptions;
  private _driver?: string;
  private _commands: Array<[string, CmdHandler]> = [];

  constructor(model: string) {
    this._model = model;
  }

  system(prompt: string): this {
    this._system = prompt;
    return this;
  }

  maxTurns(n: number): this {
    this._maxTurns = n;
    return this;
  }

  maxRetries(n: number): this {
    this._maxRetries = n;
    return this;
  }

  stream(enabled: boolean = true): this {
    this._stream = enabled;
    return this;
  }

  tool(toolDef: ToolDef): this {
    this._tools.push(toolDef);
    return this;
  }

  tools(...toolDefs: ToolDef[]): this {
    this._tools.push(...toolDefs);
    return this;
  }

  middleware(...mws: Middleware[]): this {
    this._middleware.push(...mws);
    return this;
  }

  on(event: HookEvent, callback: (...args: any[]) => any): this {
    this._hooks.push([event, callback]);
    return this;
  }

  event(eventType: StructuredEvent): this {
    this._events.push(eventType);
    return this;
  }

  events(...eventTypes: StructuredEvent[]): this {
    this._events.push(...eventTypes);
    return this;
  }

  skill(sk: Skill, config?: Record<string, any>): this {
    this._skills.push([sk, config]);
    return this;
  }

  skills(...sks: Skill[]): this {
    for (const sk of sks) {
      this._skills.push([sk]);
    }
    return this;
  }

  shell(opts?: HasShellOptions): this {
    this._shellOpts = opts ?? {};
    return this;
  }

  driver(name: string): this {
    this._driver = name;
    return this;
  }

  command(name: string, handler: CmdHandler): this {
    this._commands.push([name, handler]);
    return this;
  }

  async create(): Promise<InstanceType<any>> {
    const { StandardAgent } = await import("./standard-agent.js");
    const agent = new StandardAgent({
      model: this._model,
      system: this._system,
      maxTurns: this._maxTurns,
      maxRetries: this._maxRetries,
      stream: this._stream,
    });

    for (const td of this._tools) {
      agent.registerTool(td);
    }

    for (const mw of this._middleware) {
      agent.use(mw);
    }

    for (const [event, cb] of this._hooks) {
      agent.on(event, cb);
    }

    for (const et of this._events) {
      agent.registerEvent(et);
    }

    if (this._shellOpts !== undefined) {
      agent.initHasShell({ ...this._shellOpts, driver: this._driver });
    } else if (this._driver) {
      agent.initHasShell({ driver: this._driver });
    }

    for (const [name, handler] of this._commands) {
      agent.registerCommand(name, handler);
    }

    // Skills last (async mount)
    for (const [sk, config] of this._skills) {
      await agent.mount(sk, config);
    }

    return agent;
  }
}
